### PR TITLE
os, osd: readv

### PR DESCRIPTION
**Pull request by xiexingguo**
Here is the background for the proposal: https://github.com/ceph/ceph/pull/29588

## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard backend`
- `jenkins test docs`
- `jenkins render docs`

</details>

### CODE REVIEW

**Comment by xiexingguo** [comment]
@liewegas the test result looks fine:

http://pulpito.ceph.com/xxg-2019-08-31_09:22:11-rados:thrash-wip-sparse-read-distro-basic-smithi/

the change set is actually small (the main big part is about the read func refactor) and should be easy to review

**Comment by liewegas** [comment]
I like this much better! 

At first I thought it was weird that you have to call fiemap first... but I guess the recovery caller we care about already has that info on hand?  It also struck me as a bit funny that the result is all concatenated into a single return bufferlist.  Given that, I think ``readv`` is actually a better name for this?

**Comment by xiexingguo** [comment]
> but I guess the recovery caller we care about already has that info on hand?

yeah. actually build_push_op might call fiemap once and then split the whole fiemap into several parts for transmitting, e.g.,  due to the max-size limit of single push op.



> I think `readv` is actually a better name for this?

I agree. Repushed.

**Comment by xiexingguo** [comment]
retest this please

**Comment by xiexingguo** [comment]
http://pulpito.ceph.com/xxg-2019-09-06_06:00:01-rados:thrash-wip-sparse-read-distro-basic-smithi/
http://pulpito.ceph.com/xxg-2019-09-07_00:42:56-rados:thrash-wip-sparse-read-distro-basic-smithi/

File: src/os/bluestore/BlueStore.cc [hunk 1]
```diff
@@ -8920,60 +8823,59 @@ int BlueStore::_do_read(
...
+              pre.regs.emplace_back(region_t(pos, b_off, l, front));
+              merged = true;
+            }
+          }
+          if (!merged) {
+            read_req_t req(r_off, r_len);
+            req.regs.emplace_back(region_t(pos, b_off, l, front));
+            r2r.emplace_back(std::move(req));
+          }
+        }
```
Comment (tchaikov) [inline src/os/bluestore/BlueStore.cc]: @xiexingguo i'd suggest avoid massively reformatting source code in future.

**Comment by tchaikov** [comment]
http://pulpito.ceph.com/kchai-2019-09-08_16:10:14-rados-master-distro-basic-mira/4289804/

File: src/os/bluestore/BlueStore.cc (hunk 1 above)
Comment (xiexingguo) [reply src/os/bluestore/BlueStore.cc]: sorry, will pay attention to next time

**Pull request merged**